
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../NextflowPipeline/">
      
      
        <link rel="next" href="../api/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>ZipStrain Command Line Interface - ZipStrain</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.84d31ad4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#zipstrain-command-line-interface" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ZipStrain" class="md-header__button md-logo" aria-label="ZipStrain" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ZipStrain
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              ZipStrain Command Line Interface
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ZipStrain" class="md-nav__button md-logo" aria-label="ZipStrain" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    ZipStrain
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Welcome to ZipStrain
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tutorial
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../NextflowPipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow Pipeline for ZipStrain
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    ZipStrain Command Line Interface
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    ZipStrain Command Line Interface
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-groups" class="md-nav__link">
    <span class="md-ellipsis">
      Command Groups
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Command Groups">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-utilities-utilities" class="md-nav__link">
    <span class="md-ellipsis">
      1. Utilities (utilities)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Utilities (utilities)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#build-null-model" class="md-nav__link">
    <span class="md-ellipsis">
      Build Null Model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merge-parquet-files" class="md-nav__link">
    <span class="md-ellipsis">
      Merge Parquet Files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#process-mpileup" class="md-nav__link">
    <span class="md-ellipsis">
      Process Mpileup
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#make-bed-file" class="md-nav__link">
    <span class="md-ellipsis">
      Make BED File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-genome-lengths" class="md-nav__link">
    <span class="md-ellipsis">
      Get Genome Lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome-breadth-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Genome Breadth Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#collect-breadth-tables" class="md-nav__link">
    <span class="md-ellipsis">
      Collect Breadth Tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#strain-heterogeneity" class="md-nav__link">
    <span class="md-ellipsis">
      Strain Heterogeneity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-profile-database" class="md-nav__link">
    <span class="md-ellipsis">
      Build Profile Database
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-comparison-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Build Comparison Configuration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#to-complete-table" class="md-nav__link">
    <span class="md-ellipsis">
      To Complete Table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-gene-tools-gene_tools" class="md-nav__link">
    <span class="md-ellipsis">
      2. Gene Tools (gene_tools)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Gene Tools (gene_tools)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gene-range-table" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Range Table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-location-table" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Location Table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-compare-compare" class="md-nav__link">
    <span class="md-ellipsis">
      3. Compare (compare)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Compare (compare)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#single-genome-compare" class="md-nav__link">
    <span class="md-ellipsis">
      Single Genome Compare
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-profile-profile" class="md-nav__link">
    <span class="md-ellipsis">
      4. Profile (profile)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Profile (profile)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prepare-profiling" class="md-nav__link">
    <span class="md-ellipsis">
      Prepare Profiling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#profile-single" class="md-nav__link">
    <span class="md-ellipsis">
      Profile Single
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-run-run" class="md-nav__link">
    <span class="md-ellipsis">
      5. Run (run)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Run (run)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#profile" class="md-nav__link">
    <span class="md-ellipsis">
      Profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare-genomes" class="md-nav__link">
    <span class="md-ellipsis">
      Compare Genomes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-comparison-database" class="md-nav__link">
    <span class="md-ellipsis">
      Build Comparison Database
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#test-command" class="md-nav__link">
    <span class="md-ellipsis">
      Test Command
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Test Command">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-installation" class="md-nav__link">
    <span class="md-ellipsis">
      Test Installation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-workflows" class="md-nav__link">
    <span class="md-ellipsis">
      Example Workflows
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Example Workflows">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-profiling-and-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Profiling and Analysis Workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#large-scale-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Large-scale Analysis Workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#strain-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Strain Analysis Workflow
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-formats" class="md-nav__link">
    <span class="md-ellipsis">
      File Formats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performance-considerations" class="md-nav__link">
    <span class="md-ellipsis">
      Performance Considerations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-help" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Help
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-groups" class="md-nav__link">
    <span class="md-ellipsis">
      Command Groups
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Command Groups">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-utilities-utilities" class="md-nav__link">
    <span class="md-ellipsis">
      1. Utilities (utilities)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Utilities (utilities)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#build-null-model" class="md-nav__link">
    <span class="md-ellipsis">
      Build Null Model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merge-parquet-files" class="md-nav__link">
    <span class="md-ellipsis">
      Merge Parquet Files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#process-mpileup" class="md-nav__link">
    <span class="md-ellipsis">
      Process Mpileup
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#make-bed-file" class="md-nav__link">
    <span class="md-ellipsis">
      Make BED File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-genome-lengths" class="md-nav__link">
    <span class="md-ellipsis">
      Get Genome Lengths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genome-breadth-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Genome Breadth Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#collect-breadth-tables" class="md-nav__link">
    <span class="md-ellipsis">
      Collect Breadth Tables
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#strain-heterogeneity" class="md-nav__link">
    <span class="md-ellipsis">
      Strain Heterogeneity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-profile-database" class="md-nav__link">
    <span class="md-ellipsis">
      Build Profile Database
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-comparison-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Build Comparison Configuration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#to-complete-table" class="md-nav__link">
    <span class="md-ellipsis">
      To Complete Table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-gene-tools-gene_tools" class="md-nav__link">
    <span class="md-ellipsis">
      2. Gene Tools (gene_tools)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Gene Tools (gene_tools)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gene-range-table" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Range Table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-location-table" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Location Table
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-compare-compare" class="md-nav__link">
    <span class="md-ellipsis">
      3. Compare (compare)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Compare (compare)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#single-genome-compare" class="md-nav__link">
    <span class="md-ellipsis">
      Single Genome Compare
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-profile-profile" class="md-nav__link">
    <span class="md-ellipsis">
      4. Profile (profile)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Profile (profile)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prepare-profiling" class="md-nav__link">
    <span class="md-ellipsis">
      Prepare Profiling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#profile-single" class="md-nav__link">
    <span class="md-ellipsis">
      Profile Single
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-run-run" class="md-nav__link">
    <span class="md-ellipsis">
      5. Run (run)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Run (run)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#profile" class="md-nav__link">
    <span class="md-ellipsis">
      Profile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare-genomes" class="md-nav__link">
    <span class="md-ellipsis">
      Compare Genomes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#build-comparison-database" class="md-nav__link">
    <span class="md-ellipsis">
      Build Comparison Database
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#test-command" class="md-nav__link">
    <span class="md-ellipsis">
      Test Command
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Test Command">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#test-installation" class="md-nav__link">
    <span class="md-ellipsis">
      Test Installation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-workflows" class="md-nav__link">
    <span class="md-ellipsis">
      Example Workflows
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Example Workflows">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-profiling-and-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Profiling and Analysis Workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#large-scale-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Large-scale Analysis Workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#strain-analysis-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Strain Analysis Workflow
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-formats" class="md-nav__link">
    <span class="md-ellipsis">
      File Formats
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performance-considerations" class="md-nav__link">
    <span class="md-ellipsis">
      Performance Considerations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-help" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Help
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="zipstrain-command-line-interface">ZipStrain Command Line Interface</h1>
<p>ZipStrain provides a comprehensive command-line interface for strain-level analysis of metagenomic data. The CLI is organized into several command groups for different functionalities.</p>
<h2 id="installation">Installation</h2>
<p>First, ensure ZipStrain is installed (see <a href="../installation/">Installation.md</a> for details).</p>
<h2 id="usage">Usage</h2>
<p>The main CLI command is <code>zipstrain</code> with various subcommands organized into functional groups:</p>
<pre><code>zipstrain [GROUP] [COMMAND] [OPTIONS]
</code></pre>
<h2 id="command-groups">Command Groups</h2>
<h3 id="1-utilities-utilities">1. Utilities (<code>utilities</code>)</h3>
<p>General utility commands for data processing and preparation.</p>
<h4 id="build-null-model">Build Null Model</h4>
<p>Create a null model for sequencing errors based on statistical distributions:</p>
<pre><code>zipstrain utilities build-null-model [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-e, --error-rate FLOAT</code>: Error rate for the sequencing technology (default: 0.001)</li>
<li><code>-m, --max-total-reads INTEGER</code>: Maximum coverage to consider for a base (default: 10000)</li>
<li><code>-p, --p-threshold FLOAT</code>: Significance threshold for the Poisson distribution (default: 0.05)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output Parquet file [required]</li>
<li><code>-t, --model-type [poisson]</code>: Type of null model to build (default: poisson)</li>
</ul>
<p><strong>Example:</strong></p>
<pre><code>zipstrain utilities build-null-model -e 0.001 -m 10000 -p 0.05 -o null_model.parquet
</code></pre>
<h4 id="merge-parquet-files">Merge Parquet Files</h4>
<p>Merge multiple Parquet files into a single file:</p>
<pre><code>zipstrain utilities merge_parquet -i INPUT_DIR -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-i, --input-dir TEXT</code>: Directory containing Parquet files to merge [required]</li>
<li><code>-o, --output-file TEXT</code>: Path to save the merged Parquet file [required]</li>
</ul>
<h4 id="process-mpileup">Process Mpileup</h4>
<p>Process mpileup files and save results in Parquet format:</p>
<pre><code>zipstrain utilities process_mpileup [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-g, --gene-range-table-loc TEXT</code>: Location of the gene range table in TSV format [required]</li>
<li><code>-b, --batch-bed TEXT</code>: Location of the batch BED file [required]</li>
<li><code>-s, --batch-size INTEGER</code>: Buffer size for processing stdin from samtools (default: 10000)</li>
<li><code>-o, --output-file TEXT</code>: Location to save the output Parquet file [required]</li>
</ul>
<h4 id="make-bed-file">Make BED File</h4>
<p>Create a BED file from a FASTA database:</p>
<pre><code>zipstrain utilities make_bed -d DB_FASTA_DIR -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-d, --db-fasta-dir TEXT</code>: Path to the database in FASTA format [required]</li>
<li><code>-m, --max-scaffold-length INTEGER</code>: Maximum scaffold length to split into multiple entries (default: 500000)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output BED file [required]</li>
</ul>
<h4 id="get-genome-lengths">Get Genome Lengths</h4>
<p>Extract genome length information from scaffold-to-genome mapping:</p>
<pre><code>zipstrain utilities get_genome_lengths -s STB_FILE -b BED_FILE -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-s, --stb-file TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-b, --bed-file TEXT</code>: Path to the BED file [required]</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output Parquet file [required]</li>
</ul>
<h4 id="genome-breadth-matrix">Genome Breadth Matrix</h4>
<p>Generate a genome breadth matrix from profiles:</p>
<pre><code>zipstrain utilities genome_breadth_matrix [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-p, --profile TEXT</code>: Path to the profile Parquet file [required]</li>
<li><code>-g, --genome-length TEXT</code>: Path to the genome length Parquet file [required]</li>
<li><code>-s, --stb TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-c, --min-cov INTEGER</code>: Minimum coverage to consider a position (default: 1)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output Parquet file [required]</li>
</ul>
<h4 id="collect-breadth-tables">Collect Breadth Tables</h4>
<p>Collect multiple genome breadth tables into a single file:</p>
<pre><code>zipstrain utilities collect_breadth_tables -d BREADTH_TABLES_DIR -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-d, --breadth-tables-dir TEXT</code>: Directory containing breadth tables in Parquet format [required]</li>
<li><code>-e, --extension TEXT</code>: File extension of the breadth tables (default: parquet)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the collected breadth tables [required]</li>
</ul>
<h4 id="strain-heterogeneity">Strain Heterogeneity</h4>
<p>Calculate strain heterogeneity for each genome based on nucleotide frequencies:</p>
<pre><code>zipstrain utilities strain_heterogeneity [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-p, --profile-file TEXT</code>: Path to the profile Parquet file [required]</li>
<li><code>-s, --stb-file TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-c, --min-cov INTEGER</code>: Minimum coverage to consider a position (default: 5)</li>
<li><code>-f, --freq-threshold FLOAT</code>: Frequency threshold to define dominant nucleotide (default: 0.8)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output Parquet file [required]</li>
</ul>
<h4 id="build-profile-database">Build Profile Database</h4>
<p>Build a profile database from a CSV file:</p>
<pre><code>zipstrain utilities build-profile-db -p PROFILE_DB_CSV -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-p, --profile-db-csv TEXT</code>: Path to the profile database CSV file [required]</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output Parquet file [required]</li>
</ul>
<h4 id="build-comparison-configuration">Build Comparison Configuration</h4>
<p>Build a comparison configuration JSON file:</p>
<pre><code>zipstrain utilities build-comparison-config [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-p, --profile-db TEXT</code>: Path to the profile database Parquet file [required]</li>
<li><code>-g, --gene-db-id TEXT</code>: Gene database ID [required]</li>
<li><code>-r, --reference-db-id TEXT</code>: Reference fasta ID [required]</li>
<li><code>-s, --scope TEXT</code>: Genome scope for comparison (default: "all")</li>
<li><code>-c, --min-cov INTEGER</code>: Minimum coverage to consider a position (default: 5)</li>
<li><code>-l, --min-gene-compare-len INTEGER</code>: Minimum gene length to consider for comparison (default: 200)</li>
<li><code>-n, --null-model-p-value FLOAT</code>: P-value threshold for the null model (default: 0.05)</li>
<li><code>-t, --stb-file-loc TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-m, --null-model-loc TEXT</code>: Path to the null model Parquet file [required]</li>
<li><code>-a, --current-comp-table TEXT</code>: Path to existing comparison table in Parquet format (optional)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the output configuration JSON file [required]</li>
</ul>
<h4 id="to-complete-table">To Complete Table</h4>
<p>Generate a table of remaining genome comparison pairs:</p>
<pre><code>zipstrain utilities to-complete-table -g GENOME_COMPARISON_OBJECT -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-g, --genome-comparison-object TEXT</code>: Path to the genome comparison object in JSON format [required]</li>
<li><code>-o, --output-file TEXT</code>: Path to save the completed pairs CSV file [required]</li>
</ul>
<h3 id="2-gene-tools-gene_tools">2. Gene Tools (<code>gene_tools</code>)</h3>
<p>Commands for working with gene annotations and locations.</p>
<h4 id="gene-range-table">Gene Range Table</h4>
<p>Build a gene range table from gene files:</p>
<pre><code>zipstrain gene_tools gene-range-table -g GENE_FILE -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-g, --gene-file TEXT</code>: Location of gene file (Prodigal's nucleotide FASTA output) [required]</li>
<li><code>-o, --output-file TEXT</code>: Location to save output TSV file [required]</li>
</ul>
<h4 id="gene-location-table">Gene Location Table</h4>
<p>Build a gene location table for specified scaffolds:</p>
<pre><code>zipstrain gene_tools gene-loc-table -g GENE_FILE -s SCAFFOLD_LIST -o OUTPUT_FILE
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-g, --gene-file TEXT</code>: Location of gene file (Prodigal's nucleotide FASTA output) [required]</li>
<li><code>-s, --scaffold-list TEXT</code>: Location of scaffold list (text file with scaffold names) [required]</li>
<li><code>-o, --output-file TEXT</code>: Location to save output Parquet file [required]</li>
</ul>
<h3 id="3-compare-compare">3. Compare (<code>compare</code>)</h3>
<p>Commands for comparing genomes and samples.</p>
<h4 id="single-genome-compare">Single Genome Compare</h4>
<p>Compare two mpileup files for genome analysis:</p>
<pre><code>zipstrain compare single_compare_genome [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-m1, --mpileup-contig-1 TEXT</code>: Path to the first mpileup file [required]</li>
<li><code>-m2, --mpileup-contig-2 TEXT</code>: Path to the second mpileup file [required]</li>
<li><code>-s1, --scaffolds-1 TEXT</code>: Path to the list of scaffolds for the first mpileup file [required]</li>
<li><code>-s2, --scaffolds-2 TEXT</code>: Path to the list of scaffolds for the second mpileup file [required]</li>
<li><code>-n, --null-model TEXT</code>: Path to the null model Parquet file [required]</li>
<li><code>-s, --stb-file TEXT</code>: Path to the scaffold to genome mapping file [required]</li>
<li><code>-c, --min-cov INTEGER</code>: Minimum coverage to consider a position (default: 5)</li>
<li><code>-l, --min-gene-compare-len INTEGER</code>: Minimum gene length to consider for comparison (default: 100)</li>
<li><code>-m, --memory-mode [heavy|light]</code>: Memory mode for processing (default: heavy)</li>
<li><code>-b, --chrom-batch-size INTEGER</code>: Batch size for processing chromosomes (default: 10000)</li>
<li><code>-o, --output-file TEXT</code>: Path to save the parquet file [required]</li>
<li><code>-g, --genome TEXT</code>: If provided, do comparison only for the specified genome (default: all)</li>
<li><code>-e, --engine [streaming|gpu|auto]</code>: Engine to use for processing (default: streaming)</li>
</ul>
<h3 id="4-profile-profile">4. Profile (<code>profile</code>)</h3>
<p>Commands for profiling BAM files.</p>
<h4 id="prepare-profiling">Prepare Profiling</h4>
<p>Prepare files needed for profiling BAM files:</p>
<pre><code>zipstrain profile prepare_profiling [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-r, --reference-fasta TEXT</code>: Path to the reference genome in FASTA format [required]</li>
<li><code>-g, --gene-fasta TEXT</code>: Path to the gene annotations in FASTA format [required]</li>
<li><code>-s, --stb-file TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-o, --output-dir TEXT</code>: Directory to save the profiling database [required]</li>
</ul>
<h4 id="profile-single">Profile Single</h4>
<p>Profile a single BAM file:</p>
<pre><code>zipstrain profile profile-single [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-b, --bed-file TEXT</code>: Path to the BED file describing regions to be profiled [required]</li>
<li><code>-a, --bam-file TEXT</code>: Path to the BAM file to be profiled [required]</li>
<li><code>-g, --gene-range-table TEXT</code>: Path to the gene range table [required]</li>
<li><code>-n, --num-workers INTEGER</code>: Number of workers to use for profiling (default: 1)</li>
<li><code>-o, --output-dir TEXT</code>: Directory to save the profiling output [required]</li>
</ul>
<h3 id="5-run-run">5. Run (<code>run</code>)</h3>
<p>Commands for running large-scale analyses and managing workflows.</p>
<h4 id="profile">Profile</h4>
<p>Run BAM file profiling in batches:</p>
<pre><code>zipstrain run profile [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-i, --input-table TEXT</code>: Path to the input table in TSV format containing sample names and BAM file paths [required]</li>
<li><code>-s, --stb-file TEXT</code>: Path to the scaffold-to-genome mapping file [required]</li>
<li><code>-g, --gene-range-table TEXT</code>: Path to the gene range table file [required]</li>
<li><code>-b, --bed-file TEXT</code>: Path to the BED file for profiling regions [required]</li>
<li><code>-l, --genome-length-file TEXT</code>: Path to the genome length file [required]</li>
<li><code>-r, --run-dir TEXT</code>: Directory to save the run data [required]</li>
<li><code>-n, --num-procs INTEGER</code>: Number of processors to use for each profiling task (default: 8)</li>
<li><code>-m, --max-concurrent-batches INTEGER</code>: Maximum number of concurrent batches to run (default: 5)</li>
<li><code>-p, --poll-interval INTEGER</code>: Polling interval in seconds to check batch status (default: 1)</li>
<li><code>-e, --execution-mode TEXT</code>: Execution mode: 'local' or 'slurm' (default: local)</li>
<li><code>-c, --slurm-config TEXT</code>: Path to the SLURM configuration file in JSON format (required if execution mode is 'slurm')</li>
<li><code>-o, --container-engine TEXT</code>: Container engine to use: 'local', 'docker' or 'apptainer' (default: local)</li>
<li><code>-t, --task-per-batch INTEGER</code>: Number of tasks to include in each batch (default: 10)</li>
</ul>
<h4 id="compare-genomes">Compare Genomes</h4>
<p>Run genome comparisons in batches:</p>
<pre><code>zipstrain run compare_genomes [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-g, --genome-comparison-object TEXT</code>: Path to the genome comparison object in JSON format [required]</li>
<li><code>-r, --run-dir TEXT</code>: Directory to save the run data [required]</li>
<li><code>-m, --max-concurrent-batches INTEGER</code>: Maximum number of concurrent batches to run (default: 5)</li>
<li><code>-p, --poll-interval INTEGER</code>: Polling interval in seconds to check batch status (default: 1)</li>
<li><code>-e, --execution-mode TEXT</code>: Execution mode: 'local' or 'slurm' (default: local)</li>
<li><code>-s, --slurm-config TEXT</code>: Path to the SLURM configuration file in JSON format (required if execution mode is 'slurm')</li>
<li><code>-c, --container-engine TEXT</code>: Container engine to use: 'local', 'docker' or 'apptainer' (default: local)</li>
<li><code>-t, --task-per-batch INTEGER</code>: Number of tasks to include in each batch (default: 10)</li>
<li><code>-a, --polars-engine [streaming|gpu|auto]</code>: Polars engine to use (default: streaming)</li>
<li><code>-b, --chrom-batch-size INTEGER</code>: Batch size for processing chromosomes (default: 10000)</li>
<li><code>-h, --memory-mode [heavy|light]</code>: Memory mode for processing (default: heavy)</li>
</ul>
<h4 id="build-comparison-database">Build Comparison Database</h4>
<p>Build a genome comparison database from profiles:</p>
<pre><code>zipstrain run build-comp-database [OPTIONS]
</code></pre>
<p><strong>Options:</strong></p>
<ul>
<li><code>-p, --profile-db-dir TEXT</code>: Directory containing profile in Parquet format [required]</li>
<li><code>-c, --config-file TEXT</code>: Path to the genome comparison database config file in JSON format [required]</li>
<li><code>-o, --output-dir TEXT</code>: Directory to save genome comparison database object [required]</li>
<li><code>-f, --comp-db-file TEXT</code>: Path to initial database file (optional)</li>
</ul>
<h2 id="test-command">Test Command</h2>
<h3 id="test-installation">Test Installation</h3>
<p>Verify ZipStrain setup and dependencies:</p>
<pre><code>zipstrain test
</code></pre>
<p>This command checks if all required dependencies (like samtools) are properly installed and accessible.</p>
<h2 id="example-workflows">Example Workflows</h2>
<h3 id="basic-profiling-and-analysis-workflow">Basic Profiling and Analysis Workflow</h3>
<ol>
<li><strong>Prepare profiling files:</strong></li>
</ol>
<pre><code>zipstrain profile prepare_profiling \
  -r reference.fasta \
  -g genes.fasta \
  -s scaffold_to_genome.tsv \
  -o profiling_db
</code></pre>
<ol>
<li><strong>Build null model:</strong></li>
</ol>
<pre><code>zipstrain utilities build-null-model \
  -e 0.001 -m 10000 -p 0.05 \
  -o null_model.parquet
</code></pre>
<ol>
<li><strong>Profile single BAM file:</strong></li>
</ol>
<pre><code>zipstrain profile profile-single \
  -b profiling_db/genomes_bed_file.bed \
  -a sample1.bam \
  -g profiling_db/gene_range_table.tsv \
  -n 4 \
  -o sample1_profile
</code></pre>
<ol>
<li><strong>Compare two profiles:</strong></li>
</ol>
<pre><code>zipstrain compare single_compare_genome \
  -m1 sample1_profile/sample1.parquet \
  -m2 sample2_profile/sample2.parquet \
  -s1 sample1_profile/sample1.parquet.scaffolds \
  -s2 sample2_profile/sample2.parquet.scaffolds \
  -n null_model.parquet \
  -s scaffold_to_genome.tsv \
  -o comparison_results.parquet
</code></pre>
<h3 id="large-scale-analysis-workflow">Large-scale Analysis Workflow</h3>
<ol>
<li><strong>Build profile database:</strong></li>
</ol>
<pre><code>zipstrain utilities build-profile-db \
  -p profiles.csv \
  -o profiles.parquet
</code></pre>
<ol>
<li><strong>Build comparison configuration:</strong></li>
</ol>
<pre><code>zipstrain utilities build-comparison-config \
  -p profiles.parquet \
  -g genes_v1 \
  -r gtdb_r214 \
  -s all \
  -c 5 \
  -l 200 \
  -t scaffold_to_genome.tsv \
  -m null_model.parquet \
  -o comparison_config.json
</code></pre>
<ol>
<li><strong>Build comparison database:</strong></li>
</ol>
<pre><code>zipstrain run build-comp-database \
  -p profiles.parquet \
  -c comparison_config.json \
  -o comparison_db.json
</code></pre>
<ol>
<li><strong>Run batch comparisons:</strong></li>
</ol>
<pre><code>zipstrain run compare_genomes \
  -g comparison_db.json \
  -r run_results \
  -m 10 \
  -e slurm \
  -s slurm_config.json \
  -c apptainer
</code></pre>
<h3 id="strain-analysis-workflow">Strain Analysis Workflow</h3>
<ol>
<li><strong>Calculate strain heterogeneity:</strong></li>
</ol>
<pre><code>zipstrain utilities strain_heterogeneity \
  -p sample_profile.parquet \
  -s scaffold_to_genome.tsv \
  -c 5 \
  -f 0.8 \
  -o strain_het.parquet
</code></pre>
<ol>
<li><strong>Generate genome breadth matrix:</strong></li>
</ol>
<pre><code>zipstrain utilities genome_breadth_matrix \
  -p sample_profile.parquet \
  -g genome_lengths.parquet \
  -s scaffold_to_genome.tsv \
  -c 1 \
  -o breadth_matrix.parquet
</code></pre>
<h2 id="file-formats">File Formats</h2>
<ul>
<li><strong>Parquet</strong>: Primary format for structured data storage and processing</li>
<li><strong>TSV</strong>: Tab-separated values for simple tabular data</li>
<li><strong>BED</strong>: Standard genomic interval format</li>
<li><strong>JSON</strong>: Configuration files and object serialization</li>
<li><strong>FASTA</strong>: Sequence data input</li>
<li><strong>CSV</strong>: Simple data tables for input/output</li>
</ul>
<h2 id="performance-considerations">Performance Considerations</h2>
<ul>
<li>Use <code>--memory-mode light</code> for large datasets with limited memory</li>
<li>Choose appropriate <code>--engine</code> based on available hardware (GPU vs CPU)</li>
<li>Adjust <code>--chrom-batch-size</code> based on available memory</li>
<li>Use container engines for consistent environments across platforms</li>
<li>Consider SLURM execution mode for large-scale HPC analyses</li>
</ul>
<h2 id="getting-help">Getting Help</h2>
<p>For detailed help on any command:</p>
<pre><code>zipstrain [GROUP] [COMMAND] --help
</code></pre>
<p>For general help:</p>
<pre><code>zipstrain --help
</code></pre>
<p>To test your installation:</p>
<pre><code>zipstrain test
</code></pre>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>