# Welcome to ZipStrain

ZipStrain is a tool designed for efficient and scalable strain-level analysis of metagenomic data. It can profile and compare metagenomics samples at nucleotide resolution, providing insights into the strain composition of microbial communities.

![ZipStrain Logo](Zipstrain.svg)

## Features

- **Strain-Level Resolution Profiling and Comparison**: ZipStrain provides detailed insights into the strain composition of microbial communities, enabling researchers to identify and compare strains across different samples.

- **Scalability**: ZipStrain is built using high-performance libraries. This makes comparing samples fast. It also distributes tasks across multiple processors or nodes, making it suitable for large datasets. Finally, most the workflows are resumable.

- **Containerization**: ZipStrain is fully containerized using Docker and Singularity, ensuring consistent environments across different platforms and simplifying deployment.

- **Workflow Management**: ZipStrain utilizes Nextflow for workflow management, allowing for easy customization, reproducibility, and scalability of analyses.

- **Command-Line Interface (CLI)**: ZipStrain offers a user-friendly CLI for easy interaction and execution of tasks.

- **Python API**: ZipStrain provides a Python API for programmatic access, enabling integration into larger bioinformatics pipelines and workflows. The API also allows for downstream analyses and visualizations.

